### ISSUE TITLE
prefer-destructuring rule falsely reporting array as object when enforceForRenamedProperties is set

### ISSUE THREAD

**Issue by ismail-syed** [eslint/eslint#8654]
**Tell us about your environment**

* **ESLint Version:** 3.19.0
* **Node Version:** v7.7.3
* **npm Version:** v4..1.2

**Please show your full configuration:**
See https://github.com/Shopify/eslint-plugin-shopify
Specifically our [`prefer-destructuring`](https://github.com/Shopify/eslint-plugin-shopify/blob/master/lib/config/rules/ecmascript-6.js#L43) config

```

  'prefer-destructuring': ['error', {array: false, object: true}, {enforceForRenamedProperties: true}],

```

**What did you do? Please include the actual source code causing the issue.**

```js

const arr = [1, 2, 3];
const second = arr[1];
```

**What did you expect to happen?**
No errors, since array has been set to false

**What actually happened? Please include the actual, raw output from ESLint.**
```
  16:9  error    Use object destructuring                     prefer-destructuring
```

**Comment by ismail-syed** [comment]
It seems like the [`enforceForRenamedProperties`](https://github.com/eslint/eslint/blob/master/lib/rules/prefer-destructuring.js#L117-L119) option doesn't take into consideration the state of the array option.

**Comment by not-an-aardvark** [comment]
Thanks for the report. The `array` option refers to using array *destructuring*, like this:

```js
const [first] = arr;
```

It does not refer to whether the value that's being destructured is an array, because it's generally not possible to tell what the value of a given variable will be at runtime.

So the rule is recommending that you use this code instead:

```js
const { [1]: second } = arr;
```

**Comment by ismail-syed** [comment]
Makes sense. However, if we have array set as false, why is it still being flagged as an error?

If I remove the `enforceForRenamedProperties` option, that line doesn't get flagged as a error, which is expected, since array is set as false.

**Comment by ismail-syed** [comment]
If we have array set a true, it makes sense to raise an error and expect this instead:
```
const { [1]: second } = arr;
```

**Comment by not-an-aardvark** [comment]
Keep in mind that the rule doesn't know that `arr` is an array. It could just be a regular object that has `"1"` as a key.

**Comment by not-an-aardvark** [comment]
Although that said, if the rule assumes that accessing an integer property of an object should be array destructuring when array checking is enabled, I guess it doesn't make sense to treat it as object destructuring when array checking is disabled. I'll mark this as a bug.

**Comment by flowmemo** [comment]
I'm going to work on this issue.
Is it correct to say, the `object` option should not handle the case when accessing an integer property of an object, no matter what the `array` option is?

**Comment by not-an-aardvark** [comment]
Yes, that sounds right. Thanks for working on this!
